### ISSUE TITLE
zindex, pane, or layer ordering

### ISSUE THREAD

**Issue by dcaud** [rstudio/leaflet#427]
The current leaflet implementation doesn't seem to allow setting the zIndex or pane properly (tiles and images are placed on the tile pane whereas polygons/lines/circles are placed on the overlay pane which has a higher zIndex).

How do you set the layer order in R's leaflet package so that tiles show up on top of polygons filled with color?

More [here](http://stackoverflow.com/questions/43881007/how-to-addtiles-on-top-of-addpolygons-in-rs-leaflet?noredirect=1#comment74941092_43881007).

**Comment by jcheng5** [comment]
Can you try adding`options=list(zIndex=xxxx)` to your various layer function calls? I believe these pass straight through to leaflet.js.

**Comment by tim-salabim** [comment]
The `add*` functions don't seem to respect `zIndex`

```r
library(mapview)

leaflet() %>%
  addCircleMarkers(data = breweries, options = list(zIndex = 400), color = "red") %>%
  addPolygons(data = franconia, options = list(zIndex = 200),
              fillOpacity = 1)

leaflet() %>%
  addPolygons(data = franconia, options = list(zIndex = 400),
              fillOpacity = 1) %>%
  addCircleMarkers(data = breweries, options = list(zIndex = 200), color = "red")
```

**Comment by jcheng5** [comment]
Ohhh, I see what's happening here. I was surprised the jsfiddle in the stackoverflow post works at all--because Leaflet imposes a strict ordering between different types of objects (see http://leafletjs.com/reference-0.7.7.html#map-panes).

Look at how the jsfiddle accomplishes putting the labels above the polygons:

```javascript
//create map sandwich
var topPane = map._createPane('leaflet-top-pane', map.getPanes().mapPane);
topPane.appendChild(cdbLabels.getContainer());
```

It actually creates its own custom pane using the undocumented `_createPane`. That's why no zIndex approach alone is ever going to work.

If we wanted to support this pattern, I think it'd have to be via specifically allowing basemaps to render in a custom topmost pane, rather than trying to have a general facility for setting custom ordering across all layer types.

**Comment by tim-salabim** [comment]
Is this strict ordering also still the case in leaflet 1.0?

**Comment by jcheng5** [comment]
@tim-salabim Yeah, it would appear so: http://leafletjs.com/reference-1.0.3.html#map-pane

**Comment by tim-salabim** [comment]
@jcheng5 ok, but don't the "Options inherited from GridLayer" in http://leafletjs.com/reference-1.0.3.html#tilelayer imply that we should be able to set the pane where to render the layer?

**Comment by jcheng5** [comment]
Oh, that's news to me! It does look like not only tiles but lots of layer types (all of them?) let you specify the layer by name. So yeah, it looks like you will be able to manually put layers in whatever pane you want once we move to 1.0.x. Thanks!

**Comment by tim-salabim** [comment]
I am still scared of the move to 1.0 but this adds to easing the pain a little :-)

**Comment by dcaud** [comment]
Thanks for all the thought given to this topic!

I guess the answer at the moment is:

1. It's not currently possible to control the layer on which tiles appear in R's version of Leaflet (although workarounds are available in other implementations).

2. It may become possible in a newer version of leaflet.

Any ideas on how to know whether the new version of leaflet is available?

Thanks!

**Comment by mrmalloy** [comment]
Hi - Just wanted to second the last question on this thread. I see there have been a few new updates of Leaflet recently, but I'm assuming there's an update to the R package that comes first before the custom pane feature is accessible? Or that the R package is not yet using Leaflet 1.0.x?

Also I'm wondering, is there any way to achieve this functionality in R by adding code in an htmlwidgets::onRender statement? I'm pretty new to Leaflet/R so apologies if this is a dumb question, I don't know much about how that statement works.

**Comment by Swoorup** [comment]
It seems that leaflet-pane div has z-index set to 400 for odd reason?

**Comment by schloerke** [comment]
v1.x of leaflet.js now allows for custom panes.  @Swoorup The odd leaflet-pane z-index value is set to 400 (and other panes are on the 100's) to allow for many custom layers in between the panes.  See https://github.com/Leaflet/Leaflet/pull/3591/files

#549 addresses this issue and I believe will be the solution going forward.  Creating the custom panes allows for you to define the z-index and have more control over the plot ordering.
